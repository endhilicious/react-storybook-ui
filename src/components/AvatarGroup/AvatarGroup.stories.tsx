import { ComponentStory, ComponentMeta } from '@storybook/react';
import { MAX_LIMIT_AVATAR, MOCK_IMAGES } from '../../utils/constants';

import AvatarGroup from './AvatarGroup';

export default {
  title: 'Example/AvatarGroup',
  component: AvatarGroup,
  parameters: {
    layout: 'fullscreen',
  },
} as ComponentMeta<typeof AvatarGroup>;

const Template: ComponentStory<typeof AvatarGroup> = (args) => (
  <div style={{ padding: 14 }}>
    <AvatarGroup {...args} />
  </div>
);

export const ExtraSmall = Template.bind({});
ExtraSmall.args = {
  avatars: MOCK_IMAGES,
  size: 'xs',
  maxLength: MAX_LIMIT_AVATAR
};

export const Small = Template.bind({});
Small.args = {
  avatars: MOCK_IMAGES,
  size: 'sm',
  maxLength: MAX_LIMIT_AVATAR
};

export const Medium = Template.bind({});
Medium.args = {
  avatars: MOCK_IMAGES,
  size: 'md',
  maxLength: MAX_LIMIT_AVATAR
};

export const Large = Template.bind({});
Large.args = {
  avatars: MOCK_IMAGES,
  size: 'lg',
  maxLength: MAX_LIMIT_AVATAR
};
