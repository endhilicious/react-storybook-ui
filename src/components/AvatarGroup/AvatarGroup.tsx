import { FC, Fragment } from 'react';
import { DEFAULT_AVATAR_NAME } from '../../utils/constants';

import { Size } from '../../utils/globalTypes';
import { getAvatarTextSize, getSize } from '../../utils/helper';
import Avatar from '../Avatar/Avatar';
import styles from './styles.module.scss';

interface AvatarGroupProps {
  maxLength: number;
  size: Size;
  hideImage?: boolean;
  avatars: string[];
}

const AvatarGroup: FC<AvatarGroupProps> = ({ maxLength, size, avatars, hideImage }) => {
  const visibleAvatars = avatars.slice(0, maxLength);

  return (
    <div className={styles.AvatarGroup__wrapper} data-testid="avatar-group">
      {visibleAvatars.map((avatar, index) => (
        <Fragment key={index}>
          <Avatar
            src={avatar}
            size={size}
            classNames={styles.AvatarGroup__imageAvatar}
            hideImage={hideImage}
            avatarName={DEFAULT_AVATAR_NAME} // TAMBAHAN / TODO: customized as needed
          />
        </Fragment>
      ))}
      {avatars.length > maxLength && (
        <div
          className={styles.AvatarGroup__lists}
          style={{
            width: getSize(size),
            height: getSize(size),
          }}
        >
          <span style={{ fontSize: getAvatarTextSize(size) }}>+ {avatars.length - maxLength}</span>
        </div>
      )}
    </div>
  );
};

export default AvatarGroup;
