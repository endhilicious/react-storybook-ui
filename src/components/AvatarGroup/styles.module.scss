.AvatarGroup__wrapper {
  display: flex;
}
.AvatarGroup__lists {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;
  background-color: #E7EAF4;
  margin-left: -15px;
  border: 2px solid #fff;
}
.AvatarGroup__lists > span {
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
  color: #767B87;
}
.AvatarGroup__imageAvatar {
  border: 2px solid #fff;
  margin-left: -15px;
}
.AvatarGroup__imageAvatar:first-child {
  margin-left: 0;
}