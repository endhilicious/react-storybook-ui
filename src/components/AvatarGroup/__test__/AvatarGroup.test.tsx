import { render, screen } from '@testing-library/react';
import AvatarGroup from '../AvatarGroup';

describe('AvatarGroup', () => {
  const avatars = [
    'https://example.com/avatar1.jpg',
    'https://example.com/avatar2.jpg',
    'https://example.com/avatar3.jpg',
  ];

  it('renders the default component properly', () => {
    render(<AvatarGroup maxLength={3} size="md" avatars={avatars} />);
    const avatarGroupElement = screen.getByTestId('avatar-group');

    expect(avatarGroupElement).toBeInTheDocument();
    expect(avatarGroupElement).toHaveClass('AvatarGroup__wrapper');

    const avatarElements = screen.getAllByAltText(/avatar/);
    expect(avatarElements).toHaveLength(3);
    avatarElements.forEach((avatarElement) => {
      expect(avatarElement).toHaveClass('Avatar__image');
    });

    const avatarListElement = screen.queryByRole('div', { name: '+ 0' });
    expect(avatarListElement).not.toBeInTheDocument();
  });

  test('AvatarGroup > renders avatars according to maxLength prop', () => {
    const maxLength = 2;
    render(<AvatarGroup avatars={avatars} size="md" maxLength={maxLength} />);
  
    const avatarGroupElement = screen.getByTestId('avatar-group');
    expect(avatarGroupElement).toBeInTheDocument();
  
    const avatarImages = screen.getAllByRole('img');
    expect(avatarImages.length).toEqual(maxLength);
  
    const avatarListElement = screen.getByText(`+ ${avatars.length - maxLength}`);
    expect(avatarListElement).toBeInTheDocument();
  });
  

  it('renders the name-only component when hideImage prop is true', () => {
    const avatars = ['https://example.com/avatar1.jpg', 'https://example.com/avatar2.jpg'];
    render(<AvatarGroup maxLength={2} size="md" avatars={avatars} hideImage />);
    const nameOnlyElement = screen.queryByTestId('avatar-group');
    expect(nameOnlyElement).toBeInTheDocument();
    expect(nameOnlyElement).toHaveClass('AvatarGroup__wrapper');
    expect(screen.queryByText('+ 1')).toBeNull();
  });

  it('matches snapshot', () => {
    const { container } = render(<AvatarGroup maxLength={2} size="md" avatars={avatars} />);
    expect(container).toMatchSnapshot();
  });
});
