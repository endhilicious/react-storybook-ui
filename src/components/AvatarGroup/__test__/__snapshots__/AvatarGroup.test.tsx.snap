// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup matches snapshot 1`] = `
<div>
  <div
    class="AvatarGroup__wrapper"
    data-testid="avatar-group"
  >
    <img
      alt="https://example.com/avatar1.jpg"
      class="Avatar__image AvatarGroup__imageAvatar"
      height="48"
      src="https://example.com/avatar1.jpg"
      width="48"
    />
    <img
      alt="https://example.com/avatar2.jpg"
      class="Avatar__image AvatarGroup__imageAvatar"
      height="48"
      src="https://example.com/avatar2.jpg"
      width="48"
    />
    <div
      class="AvatarGroup__lists"
      style="width: 48px; height: 48px;"
    >
      <span
        style="font-size: 20px;"
      >
        + 
        1
      </span>
    </div>
  </div>
</div>
`;

exports[`AvatarGroup matches the component snapshot 1`] = `
<div>
  <div
    class="AvatarGroup__wrapper"
    data-testid="avatar-group"
  >
    <img
      alt="https://testing123123.com/avatar1.jpg"
      class="Avatar__image AvatarGroup__imageAvatar"
      height="48"
      src="https://testing123123.com/avatar1.jpg"
      width="48"
    />
    <img
      alt="https://testing123123.com/avatar2.jpg"
      class="Avatar__image AvatarGroup__imageAvatar"
      height="48"
      src="https://testing123123.com/avatar2.jpg"
      width="48"
    />
    <img
      alt="https://testing123123.com/avatar3.jpg"
      class="Avatar__image AvatarGroup__imageAvatar"
      height="48"
      src="https://testing123123.com/avatar3.jpg"
      width="48"
    />
    <div
      class="AvatarGroup__lists"
      style="width: 48px; height: 48px;"
    >
      + 
      2
    </div>
  </div>
</div>
`;
